controllers/dockermachine_controller.go

```diff
@@ -115,6 +118,38 @@ func (r *DockerMachineReconciler) Reconcile(req ctrl.Request) (ctrl.Result, erro
 		return ctrl.Result{}, err
 	}
 
+
+	// If the DockerMachine doesn't have finalizer, add it.
+	if !util.Contains(dockerMachine.Finalizers, capiv1alpha2.MachineFinalizer) {
+		dockerMachine.Finalizers = append(dockerMachine.Finalizers, capiv1alpha2.MachineFinalizer)
+		if err := r.Client.Patch(ctx, dockerMachine, patch) ; err!= nil {
+
+			return ctrl.Result{}, errors.Wrapf(err, "failed to add finalizer to DockerMachine %q in namespace %q",
+				dockerMachine.Name, dockerMachine.Namespace)
+		}
+
+		// Since adding the finalizer updates the object return to avoid later update issues
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	//reconcileDelete dockerMachine
+	if !machine.ObjectMeta.DeletionTimestamp.IsZero() {
+		log.Info("Start reconcileDelete dockerMachine")
+		result, err := r.reconcileDelete(ctx, cluster, machine, dockerMachine)
```

Comment:
If you use `defer` for the patching, this can be simplified to just `return r.reconcileDelete(ctx, cluster, machine, dockerMachine)`